### ISSUE TITLE
Requesting an `on recording finished` block (or similar)

### ISSUE THREAD

**Issue by kiki-lee** [microsoft/pxt-microbit#5067]
As an educator, one of the first scenarios that comes to mind with the new record features is the "talking parrot" idea.  The micro:bit listens for a sound from the user, then plays it back to them.  

What I expect to be able to write as:
![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/1633958/233705797-8238e652-dde2-4392-9c19-310fb6fa16c5.png)
doesn't work as expected, because the play feature automatically stops the record feature.  I didn't recognize this when I tried it on my computer, since the simulator doesn't actually record/play.

As a teacher, I would think something is broken. After debuggig, I discover that play automatically stops the record, so I look for an `on recording finished` event.

In the absence of one, I recognize that I need to run a check for a "full" recording inside of a forever loop, but the forever loop registers the buffer as "full" as soon as it's not empty, so it again starts playback immediately, cancelling the recording. 

When I add another check to be sure it's not still recording before it begins to play, I get a full recording...but the sound plays over and over forever.

I have to go one step further and add a flag that flips after it's been played in order to get it to only fire one time...but now I can't record again. So I need to reset the flag after the A button is pressed. 

I got to what I need, but my program looks like this:
![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/1633958/233708091-07358e81-7143-4b05-9ef8-9e0001ae05a3.png)

Instead of this:
![image](https://user-images.githubusercontent.com/1633958/233711937-9c86d00e-3b0f-4c8f-8e73-d3d458e3bb07.png)

or this:
![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/1633958/233705797-8238e652-dde2-4392-9c19-310fb6fa16c5.png)

**Comment by abchatra** [comment]
@thsparks can you add your use case here as well (in case if it is not the same as above).

**Comment by thsparks** [comment]
My scenarios were similar to Kiki's but did have a few differences, so I'll describe it below. I think it could've been simplified considerably with just a few additional blocks (or tweaks to existing ones).

# Scenario
The goal was simply: When the user presses A, start recording and show an "R" on the LED screen. Once it's done recording, show a check mark on the screen. Then, when the user presses "B", show a Play icon on the screen and play the audio. Once the audio is done playing, clear the screen. The code to do this felt really unnecessarily complicated:

<img width="598" alt="image" src="https://user-images.githubusercontent.com/69657545/233737206-904bd4c8-7bba-417b-90d4-be0a7b1a0227.png">

(I would provide a share link, but I don't think it will work because it was created in beta).

# Details
Notice how it required me to create two separate variables to track if recording has started and if playing has started, then check those variables inside of a forever loop while setting the screen content, so we don't continually overwrite the screen content when we don't mean to do so.

Another note that I could see easily tripping up students (it certainly confused me at first): the order of the calls inside `on A button pressed` is actually very important. If you start the recording after setting `recordingStarted` to 1, it breaks, because between the calls to `set recordingStarted to 1` and `start recording`, the `forever` loop actually detects `recordingStarted == 1` and switches to the checkmark, all before the recording is actually started.

# Recommended Adjustments
## Make the record audio block blocking (or perhaps give it an optional "in the background" / "and wait until done" flag).
I think this would also help address some of Kiki's concerns above. If this were done, then I would not need to track `recordingStarted` or worry about it inside the `forever` loop. I'd just have three blocks in `on A Pressed` (optionally 4, including one to set sensitivity at the top):
<img width="197" alt="image" src="https://user-images.githubusercontent.com/69657545/233738381-a7cdfba5-a2f1-41f8-9c41-1135550d5b05.png">

This could also be done with an `on recording finished` event block, but I think blocking/non-blocking options feels slightly cleaner. I think either one would be an improvement, though.

## Make the play audio block blocking, or with an optional "in the background" flag.
With this, I could remove the need for `audioPlayingStarted`, and I could take all of that outside of the `forever` loop as well. It would just be:
<img width="123" alt="image" src="https://user-images.githubusercontent.com/69657545/233739020-79188286-9823-460b-933e-7aad4b4bc104.png">

Similarly to above, this could also be done with an `on recording done playing` type of block, though I have a slight preference for making the existing block optionally blocking rather than adding a new event block. 

# Summary
So with just those two changes above, the code goes from this:
<img width="598" alt="image" src="https://user-images.githubusercontent.com/69657545/233737206-904bd4c8-7bba-417b-90d4-be0a7b1a0227.png">

To this:
<img width="327" alt="image" src="https://user-images.githubusercontent.com/69657545/233738948-9d74e0ef-d9a3-4cb2-afda-73d94042ebcd.png">

Which I think would be a considerable improvement for a scenario that feels (to me) to be pretty main-line.

I had a more complex scenario where I was changing the sensitivity when the user pressed the microbit logo, and that was even worse because we have no way to "get" the sensitivity, so I needed yet another variable to track that, but I think it's okay to consider programmatically setting sensitivity to be an edge case, so maybe not as important.

**Comment by abchatra** [comment]
This is a request to make "record audio" & "play audio" to be blocking by default and optional background parameter for not blocking scenarios.

@jaustin @JohnVidler @srietkerk let me know your thoughts.

**Comment by jaustin** [comment]
I'm not sure what to think! I was originally definitely in favour of a (+) on the 'record' block with a dropdown, but I think those are ultimately quite confusing and not that well used. But on the other hand, the MicroPython function to show an image is non-blocking and it definitely confuses people initially (compared to the default 400ms pause in many MakeCode blocks...)

One thing that I do think is worth sharing is that in both cases listed above, I think there are much simpler solutions without changing the blocks.

For the parrot (and really for the other case too, I think!), just a pause works well. For the display of record/play icons you can use a busy loop too. @kiki-lee @thsparks WDYT about these versions (albeit combined into one program for my testing!)?
![image](https://user-images.githubusercontent.com/147113/233750880-7347e340-5c80-4449-91f6-62f4edb4a4a7.png)

...but then busy loops aren't that common in MakeCode so it's a bit odd...

**Comment by jaustin** [comment]
On reflection I think me describing the (+) on a block as...
> ultimately quite confusing and not that well used.

...is probably unfair. It's more that they are complexity that would be best avoided if we don't need it. I think really they are actually a pretty good way to let someone looking for a different way of doing things find it without introducing extra blocks or adding too much immediate complexity to simple blocks.

Do we have telemetry on the (+) on any popular blocks?

**Comment by thsparks** [comment]
I think a (+) still feels like the least confusing way to handle these kinds of scenarios. What Kiki & I did and the busy loops all feel like they'd be far more complex than just adding that expandable field.  I'm not sure on telemetry for the (+) off the top of my head, but what questions are you trying to answer for it? Just the usage, or something else?

**Comment by abchatra** [comment]
We will try to mimic this block
![image](https://user-images.githubusercontent.com/6107272/235265817-b239b95d-24af-458b-bae1-ffa5328247d2.png)

Everyone agrees that by default blocking (a.k.a until done) is better.

**Comment by srietkerk** [comment]
Just getting around to reading through all of this. 

I agree, recording and playback should reflect the `play sound` block. However, I think having those descriptors by default are helpful to understanding what happens when you use the block, so I'm not planning to hide those options by an expandable field. The block will look like this:
<img width="202" alt="image" src="https://user-images.githubusercontent.com/49178322/235995557-15c7dd05-97c9-45fb-b3e5-b0a63d6560f4.png">

However, I do still see value in an `on recording finished` block. The following event block was in the original `codal-audio` extension. 

<img width="169" alt="image" src="https://user-images.githubusercontent.com/49178322/235986892-f8fa958c-3495-4d16-ab38-06939eeb1027.png">

I ended up removing it, the logic being that you can achieve similar checks with the booleans, as illustrated by @jaustin's great example here
 ![image](https://user-images.githubusercontent.com/147113/233750880-7347e340-5c80-4449-91f6-62f4edb4a4a7.png)
> 
> ...but then busy loops aren't that common in MakeCode so it's a bit odd...

However, if busy loops aren't common in MakeCode, even though I think this example is really nice, I think we should have blocks that reflect our patterns. I would like to add the `on recording` block back into the set of blocks. 

In summary, I'm going to change `record audio clip` and `play audio clip` to have the `until done` or `in background` dropdown and I want to add the event block back.

Please let me know what you think @kiki-lee @abchatra @jaustin @thsparks

**Comment by abchatra** [comment]
Is on recording block unrelated to this issue? If yes let's file a separate issue and discuss there and prioritze .

There is conclusion on this issue for making the record audio clip & play audio clip blocking with option of background.

**Comment by thsparks** [comment]
Personally, I think if `record audio clip` and `play audio clip` are optionally blocking, then the need for an `on recording` block is much less significant. There may be some edge cases where it's still helpful to have them, but I think it's much less pressing since you can just take whatever you wanted inside `on recording finished` and put it after the `record audio clip` block. You wouldn't need the busy loops anymore because the blocks above them would just block, so that wouldn't be a problem.

**Comment by srietkerk** [comment]
> Personally, I think if `record audio clip` and `play audio clip` are optionally blocking, then the need for an `on recording` block is much less significant. There may be some edge cases where it's still helpful to have them, but I think it's much less pressing since you can just take whatever you wanted inside `on recording finished` and put it after the `record audio clip` block. You wouldn't need the busy loops anymore because the blocks above them would just block, so that wouldn't be a problem.

That makes sense, thanks Thomas. I'll just change it to the drop down as discussed.
